File: 3Sum/3Sum.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
int** threeSum(int* nums, int numsSize, int* returnSize, int** returnColumnSizes) {

    // amount of triplets found
    *returnSize = 0;

    // amount of numbers in every triplet (typically 3, sometimes NULL)
    *returnColumnSizes = NULL;

    // sort the array first
    qsort(nums, numsSize, sizeof(int), compare);

    int** result = (int**)malloc(numsSize * numsSize *sizeof(int*));
    *returnColumnSizes = (int*)malloc(numsSize * numsSize * sizeof(int));

    // set the first number
    for (int i = 0; i < numsSize-2; i++) {

        // skip duplicates for the first number
        if (i > 0 && nums[i] == nums[i - 1]) {
            continue;
        }

        int target = -nums[i];
        int left = i + 1, right = numsSize - 1;

        // find the remaining two numbers
        while (left < right) {
            int sum = nums[left] + nums[right];

            if (sum == target) {

                // store the triplet in sorted order
                result[*returnSize] = (int*)malloc(sizeof(int) * 3);
                result[*returnSize][0] = nums[i];
                result[*returnSize][1] = nums[left];
                result[*returnSize][2] = nums[right];
                (*returnColumnSizes)[*returnSize] = 3;
                (*returnSize)++;

                // skip duplicates for the second and third numbers
                while (left < right && nums[left] == nums[left + 1]) left++;
                while (left < right && nums[right] == nums[right - 1]) right--;

                // move pointers inward
                left++;
                right--;
            } else if (sum < target) {
                left++;
            } else {
                right--;
            }
        }
    }

    if (*returnSize > 0) {
        return result;
    } else {
        // no triplets found
        free(result);
        free(*returnColumnSizes);
        *returnColumnSizes = NULL;
        return NULL;
    }
}

int main() {
    return 1;
}

File: 3Sum/3sum.py
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:

        solutions = []
        hashmap = {}

        for i, num in enumerate(nums) :
            difference = target - num

            if difference in hashmap :
                solutions.append([difference, num])
            
            hashmap[num] = i

        return solutions


    def threeSum(self, nums: list[int]) -> list[list[int]]:

        solutions = []
        nums.sort()
        
        # do the two-sum problem, but as first number take any single one from the list
        for i, firstNum in enumerate(nums) :

            if i > 0 and nums[i] == nums[i-1] :
                continue

            target = -firstNum
            twoSum_nums = nums[i+1:]

            twoSum_answer = self.twoSum(twoSum_nums, target)
            if twoSum_answer != [] :
                for pair in twoSum_answer:
                    triplet = [firstNum] + pair
                    if triplet not in solutions:
                        solutions.append(triplet) 

        return solutions


    

        

print(Solution().threeSum([-1,0,1,2,-1,-4]), "\n")
print(Solution().threeSum([0,1,1]), "\n")
print(Solution().threeSum([0,0,0]))
print(Solution().threeSum([0,0,0,0]))

